crates/astria-sequencer/src/fees/mod.rs

```diff
@@ -81,281 +79,378 @@ impl Fee {
 }
 
 #[async_trait::async_trait]
-impl FeeHandler for Transfer {
-    #[instrument(skip_all, err)]
-    async fn check_and_pay_fees<S: StateWrite>(&self, state: S) -> eyre::Result<()> {
+pub(crate) trait FeeHandler: Send {
+    /// The Pascal-case type name, e.g. `RollupDataSubmission`.
+    // NOTE: We only require this function due to `IbcRelay` not implementing `Protobuf`.
+    fn name() -> &'static str;
+
+    /// The full name including the protobuf package, e.g.
+    /// `astria.protocol.transaction.v1.RollupDataSubmission`.
+    // NOTE: We only require this function due to `IbcRelay` not implementing `Protobuf`.
+    fn full_name() -> String;
+
+    /// The snake-case type name, e.g. `rollup_data_submission`.
+    fn snake_case_name() -> &'static str;
+
+    fn variable_component(&self) -> u128;
+
+    fn fee_asset(&self) -> Option<&asset::Denom>;
+
+    #[instrument(skip_all, err(level = Level::WARN))]
+    async fn check_and_pay_fees<'a, S>(&self, mut state: S) -> eyre::Result<()>
+    where
+        S: StateWrite,
+        FeeComponents<Self>: TryFrom<StoredValue<'a>, Error = Report>,
+    {
         let fees = state
-            .get_transfer_fees()
+            .get_fees::<Self>()
+            .await
+            .wrap_err_with(|| format!("error fetching {} fees", Self::name()))?
+            .ok_or_else(|| {
+                eyre!(
+                    "{} fees not found, so this action is disabled",
+                    Self::name()
+                )
+            })?;
```

Comment:
nit: not sure if it'll even end up saving any lines since you might still have to use `format!`, but I believe `eyre::OptionExt::ok_or_eyre` provides this functionality in a slightly more compact way

Reply:
`eyre::OptionExt::ok_or_eyre` is designed to be used with static errors, whereas `Option::ok_or_else` takes a closure which is lazily executed, i.e. it's only executed if the `Option` is `None`, and we avoid the cost of constructing a new `String` here when the `Option` is `Some`.